File: packages/sdk-typescript/tests/unit/profile-snapshots.test.ts
// V-376 — ProfileSnapshotsResource unit tests.

import { describe, expect, it, vi } from 'vitest';
import {
  ProfileSnapshotsResource,
  type ProfileSnapshotsListPage,
} from '../../src/resources/profile-snapshots.js';
import type { HttpClient } from '../../src/http.js';
import type { Profile, ProfileSnapshot } from '@driftstack/api-types';

function fakeSnapshot(id: string, parentId: string | null = 'prof_p'): ProfileSnapshot {
  return {
    id,
    parent_profile_id: parentId,
    label: `label-${id}`,
    description: null,
    parent_archetype: 'iphone16pro_ios18_7_safari26_4',
    parent_name: 'parent',
    captured_at: '2026-05-09T00:00:00Z',
    created_at: '2026-05-09T00:00:00Z',
  };
}

interface RequestOpts {
  method: string;
  path: string;
  query?: Record<string, unknown>;
  body?: unknown;
}

describe('ProfileSnapshotsResource', () => {
  it('capture POSTs to the per-profile path with the body verbatim', async () => {
    const seen: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      seen.push(opts);
      return Promise.resolve(fakeSnapshot('psnap_1'));
    });
    const http = { request } as unknown as HttpClient;
    const r = new ProfileSnapshotsResource(http);
    await r.capture('prof_p', { label: 'before-migration', description: 'pre-iOS-18' });
    expect(seen[0]).toEqual({
      method: 'POST',
      path: '/v1/profiles/prof_p/snapshots',
      body: { label: 'before-migration', description: 'pre-iOS-18' },
    });
  });

  it('list calls the cross-account path; listForProfile narrows to per-profile', async () => {
    const seen: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      seen.push(opts);
      return Promise.resolve({ data: [], has_more: false, next_cursor: null });
    });
    const http = { request } as unknown as HttpClient;
    const r = new ProfileSnapshotsResource(http);
    await r.list({ limit: 5 });
    await r.listForProfile('prof_p', { limit: 5, cursor: 'cur_x' });
    expect(seen).toEqual([
      { method: 'GET', path: '/v1/profile-snapshots', query: { limit: 5 } },
      {
        method: 'GET',
        path: '/v1/profiles/prof_p/snapshots',
        query: { limit: 5, cursor: 'cur_x' },
      },
    ]);
  });

  it('iterate walks all pages via the cursor helper', async () => {
    const seenQueries: Array<Record<string, unknown>> = [];
    const responses: ProfileSnapshotsListPage[] = [
      {
        data: [fakeSnapshot('psnap_1'), fakeSnapshot('psnap_2')],
        has_more: true,
        next_cursor: 'cur_2',
      },
      { data: [fakeSnapshot('psnap_3')], has_more: false, next_cursor: null },
    ];
    let i = 0;
    const request = vi.fn((opts: RequestOpts) => {
      seenQueries.push(opts.query ?? {});
      const resp = responses[i]!;
      i += 1;
      return Promise.resolve(resp);
    });
    const http = { request } as unknown as HttpClient;
    const r = new ProfileSnapshotsResource(http);
    const ids: string[] = [];
    for await (const s of r.iterate({ limit: 2 })) {
      ids.push(s.id);
    }
    expect(ids).toEqual(['psnap_1', 'psnap_2', 'psnap_3']);
    expect(seenQueries).toEqual([{ limit: 2 }, { limit: 2, cursor: 'cur_2' }]);
  });

  it('restore POSTs the new name and returns a Profile', async () => {
    const seen: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      seen.push(opts);
      return Promise.resolve({
        id: 'prof_new',
        name: 'restored-baseline',
        archetype: 'iphone16pro_ios18_7_safari26_4',
        description: null,
        last_used_at: null,
        created_at: '2026-05-09T00:00:00Z',
        updated_at: '2026-05-09T00:00:00Z',
      } as Profile);
    });
    const http = { request } as unknown as HttpClient;
    const r = new ProfileSnapshotsResource(http);
    const out = await r.restore('psnap_1', { name: 'restored-baseline' });
    expect(out.id).toBe('prof_new');
    expect(seen[0]).toEqual({
      method: 'POST',
      path: '/v1/profile-snapshots/psnap_1/restore',
      body: { name: 'restored-baseline' },
    });
  });

  it('delete sends DELETE to the snapshot path', async () => {
    const seen: RequestOpts[] = [];
    const request = vi.fn((opts: RequestOpts) => {
      seen.push(opts);
      return Promise.resolve(undefined);
    });
    const http = { request } as unknown as HttpClient;
    const r = new ProfileSnapshotsResource(http);
    await r.delete('psnap_1');
    expect(seen[0]).toEqual({ method: 'DELETE', path: '/v1/profile-snapshots/psnap_1' });
  });
});

File: packages/sdk-go/profile_snapshots_test.go
package driftstack

import (
	"context"
	"encoding/json"
	"net/http"
	"testing"
	"time"
)

func snapshotFixture(id string) ProfileSnapshot {
	return ProfileSnapshot{
		ID:              id,
		ParentProfileID: stringPtr("prof_p"),
		Label:           "label-" + id,
		ParentArchetype: "iphone16pro_ios18_7_safari26_4",
		ParentName:      "parent",
		CapturedAt:      time.Date(2026, 5, 9, 0, 0, 0, 0, time.UTC),
		CreatedAt:       time.Date(2026, 5, 9, 0, 0, 0, 0, time.UTC),
	}
}

func TestProfileSnapshots_Capture(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/profiles/prof_p/snapshots" || r.Method != "POST" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		var body CaptureSnapshotRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body.Label != "before-iOS-26" {
			t.Errorf("label=%q", body.Label)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(snapshotFixture("psnap_1"))
	})
	got, err := client.ProfileSnapshots.Capture(
		context.Background(),
		"prof_p",
		&CaptureSnapshotRequest{Label: "before-iOS-26"},
	)
	if err != nil {
		t.Fatal(err)
	}
	if got.ID != "psnap_1" {
		t.Errorf("id=%q", got.ID)
	}
}

func TestProfileSnapshots_ListForProfile(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/profiles/prof_p/snapshots" || r.Method != "GET" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(ProfileSnapshotsListPage{
			Data:    []ProfileSnapshot{snapshotFixture("psnap_1")},
			HasMore: false,
		})
	})
	got, err := client.ProfileSnapshots.ListForProfile(
		context.Background(),
		"prof_p",
		&ListProfileSnapshotsQuery{Limit: 10},
	)
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Data) != 1 {
		t.Errorf("got %d snapshots", len(got.Data))
	}
}

func TestProfileSnapshots_List_CrossAccount(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/profile-snapshots" || r.Method != "GET" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(ProfileSnapshotsListPage{
			Data:    []ProfileSnapshot{snapshotFixture("psnap_a"), snapshotFixture("psnap_b")},
			HasMore: false,
		})
	})
	got, err := client.ProfileSnapshots.List(
		context.Background(),
		&ListProfileSnapshotsQuery{Limit: 25},
	)
	if err != nil {
		t.Fatal(err)
	}
	if len(got.Data) != 2 {
		t.Errorf("got %d snapshots", len(got.Data))
	}
}

func TestProfileSnapshots_Iterate_WalksCursorPages(t *testing.T) {
	t.Parallel()
	pageOne := ProfileSnapshotsListPage{
		Data: []ProfileSnapshot{
			snapshotFixture("psnap_1"),
			snapshotFixture("psnap_2"),
		},
		HasMore:    true,
		NextCursor: stringPtr("psnap_2"),
	}
	pageTwo := ProfileSnapshotsListPage{
		Data:    []ProfileSnapshot{snapshotFixture("psnap_3")},
		HasMore: false,
	}
	requestN := 0
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		requestN++
		w.Header().Set("content-type", "application/json")
		if r.URL.Query().Get("cursor") == "" {
			_ = json.NewEncoder(w).Encode(pageOne)
		} else {
			_ = json.NewEncoder(w).Encode(pageTwo)
		}
	})
	var seen []string
	err := client.ProfileSnapshots.Iterate(
		context.Background(),
		nil,
		func(s *ProfileSnapshot) (bool, error) {
			seen = append(seen, s.ID)
			return true, nil
		},
	)
	if err != nil {
		t.Fatal(err)
	}
	if len(seen) != 3 {
		t.Errorf("seen %d snapshots; want 3 (got %v)", len(seen), seen)
	}
	if requestN != 2 {
		t.Errorf("requests %d; want 2", requestN)
	}
}

func TestProfileSnapshots_Restore(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/profile-snapshots/psnap_1/restore" || r.Method != "POST" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		var body RestoreSnapshotRequest
		if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
			t.Fatal(err)
		}
		if body.Name != "restored-baseline" {
			t.Errorf("name=%q", body.Name)
		}
		w.Header().Set("content-type", "application/json")
		_ = json.NewEncoder(w).Encode(profileFixture("prof_new"))
	})
	got, err := client.ProfileSnapshots.Restore(
		context.Background(),
		"psnap_1",
		&RestoreSnapshotRequest{Name: "restored-baseline"},
	)
	if err != nil {
		t.Fatal(err)
	}
	if got.ID != "prof_new" {
		t.Errorf("id=%q", got.ID)
	}
}

func TestProfileSnapshots_Delete(t *testing.T) {
	t.Parallel()
	_, client := newServer(t, func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/profile-snapshots/psnap_1" || r.Method != "DELETE" {
			t.Errorf("unexpected: %s %s", r.Method, r.URL.Path)
		}
		w.WriteHeader(204)
	})
	if err := client.ProfileSnapshots.Delete(context.Background(), "psnap_1"); err != nil {
		t.Errorf("err=%v", err)
	}
}
